### ISSUE TITLE
connection and cursor in a with block

### ISSUE THREAD

**Issue by mikehinchey** [omnisci/pymapd#234]
I wanted to use a pymapd connection in a python `with` block, but found that `__enter__` opens a cursor.  It would be useful if it returned `self`.  That way, I could write:

```python
with pymap.connect(...) as c:
    print(type(c))
    print(c.get_tables())
```
The above fails because `c` is a cursor so doesn't have `get_tables`.

This also fails because `Cursor.__enter__` does `pass` instead of `return self`.

```python
with pymap.connect(...).cursor() as c:
    print(type(c))
    print(c.execute('select count(*) from flights').fetchall())
```
 https://github.com/omnisci/pymapd/blob/0ffc1afd42aa4aa13550580a5065f0baa86ad372/pymapd/connection.py#L217

https://github.com/omnisci/pymapd/blob/0ffc1afd42aa4aa13550580a5065f0baa86ad372/pymapd/cursor.py#L26

Let me know if you agree and I'll submit a patch.

**Comment by randyzwitch** [comment]
I believe the standard for context managers specifies that it's a cursor, so I'm not sure we can change this and still be DBI compliant. Let me search a bit.

**Comment by randyzwitch** [comment]
PEP 343 is the best reference I can find so far: https://docs.python.org/2.5/whatsnew/pep-343.html

> The __enter__() method is pretty easy, having only to start a new transaction. For this application the resulting cursor object would be a useful result, so the method will return it. The user can then add as cursor to their 'with' statement to bind the cursor to a variable name.
> 
> class DatabaseConnection:
>     ...
>     def __enter__ (self):
>         # Code to start a new transaction
>         cursor = self.cursor()
>         return cursor

I get the feeling that either DBI or SQLAlchemy expects that `__enter__()` into a `Connection` returns a `Cursor`, since the pattern would be running some statements and committing them if they are successful. It's a tough call here, because yours is a valid use case, but I suspect methods such as `pandas.read_sql` (which takes a `Connection` object) is possibly looking for a cursor.

Perhaps the best thing would be to try to change this to the behavior you want, then see if you can do the following:

```python
from pymapd import connect
import pandas as pd

conn = connect(...)
df = pd.read_sql("select * from table", conn)
```
If that pandas code still works, then I'd be okay with changing the behavior (though, we should ask Quansight as well to understand what Ibis looks for)

**Comment by randyzwitch** [comment]
Closed by #237
